lib/UndoInstCombinePass.cpp

```diff
@@ -66,13 +67,35 @@ bool clspv::UndoInstCombinePass::runOnFunction(Function &F) {
   return changed;
 }
 
+VectorType *InferTypeForOpaqueLoad(LoadInst *load, uint64_t vec_size) {
+  assert(load);
+  assert(load->getPointerOperandTy()->isOpaquePointerTy());
+  // Calculate the largest vector that the target can be split into.
+  // If the indexes don't work at the resulting size then they won't work for a
+  // smaller size.
+  uint64_t new_size = clspv::SPIRVMaxVectorSize();
+  while (vec_size % new_size) { // will always break at new_size == 1
+    new_size -= 1;
+  }
+  uint64_t divisor = vec_size / new_size;
+
+  const auto old_type = llvm::cast<VectorType>(load->getType());
+  auto new_bit_width =
+      cast<IntegerType>(old_type->getElementType())->getBitWidth() * divisor;
+
+  return VectorType::get(
+      IntegerType::get(old_type->getContext(), new_bit_width), new_size,
+      old_type->getElementCount().isScalable());
+}
+
 bool clspv::UndoInstCombinePass::UndoWideVectorExtractCast(Instruction *inst) {
   auto extract = dyn_cast<ExtractElementInst>(inst);
   if (!extract)
     return false;
 
   auto vec_ty = extract->getVectorOperandType();
-  if (vec_ty->getElementCount().getKnownMinValue() <= 4)
+  auto vec_size = vec_ty->getElementCount().getKnownMinValue();
+  if (vec_size <= clspv::SPIRVMaxVectorSize())
```

Comment:
This change causes the following tests to fail:
* UndoInstCombine/undo_extract_cast.cl
* UndoInstCombine/undo_extract_cast_2xi64_to_8xi16.ll
* UndoInstCombine/undo_extract_cast_4xi32_to_16xi8.ll

This .ll files are hard-coded to only run the undo-instcombine passes, so this is expected, the undo_extract_cast.cl fail is more interesting. The LongVectorLoweringPass only seems to convert vectors with 8 or 16 elements in the cases of call instructions, cast instructions, shuffle vector instructions, function parameters or function return values.
Do you think extractelement instructions were intensionally left out or is this a bug?

Reply:
Actually, I suppose in this case we expect there to be a cast to the size 16 vector, so the LongVectorLowering pass should pick that up. I will look into it some more.

Reply:
@alan-baker The reason that LongVectorLowering isn't working in this case is that is it run before the standard LLVM passes (including InstCombine passes) while UndoInstCombine is run after. This is also likely why the size of vector to be transformed was hardcoded to over 4, since larger vectors would have already been converted to arrays.

Potentially, another pass of LongVectorLowering could be run after the standard LLVM passes, but I think it would be easier to just go back to working on all vectors over the size of 4.

Reply:
That makes sense. Thanks for digging into it.